qiskit/quantum_info/synthesis/qsd.py

```diff
@@ -0,0 +1,214 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Quantum Shannon Decomposition.
+
+Method is described in arXiv:quant-ph/0406176.
+"""
+import scipy
+import numpy as np
+from qiskit.circuit import QuantumCircuit, QuantumRegister
+from qiskit.dagcircuit.dagnode import DAGOutNode
+from qiskit.quantum_info.synthesis import two_qubit_decompose, one_qubit_decompose
+from qiskit.converters import dag_to_circuit, circuit_to_dag
+from qiskit.circuit.library.standard_gates import CZGate, CXGate
+from qiskit.quantum_info.operators.predicates import is_hermitian_matrix
+
+
+class QuantumShannonDecomposer:
+    """Class representation of Quantum Shannon Decomposition."""
+
+    def __call__(self, unitary_matrix, opt_a1=True, opt_a2=False):
+        return qs_decomposition(unitary_matrix, opt_a1=opt_a1, opt_a2=opt_a2)
+
+
+def qs_decomposition(mat, opt_a1=True, opt_a2=False, decomposer_1q=None, decomposer_2q=None):
+    """
+    Decomposes unitary matrix into one and two qubit gates using Quantum Shannon Decomposition.
+
+       ┌───┐               ┌───┐     ┌───┐     ┌───┐
+      ─┤   ├─       ───────┤ Rz├─────┤ Ry├─────┤ Rz├─────
+       │   │    ≃     ┌───┐└─┬─┘┌───┐└─┬─┘┌───┐└─┬─┘┌───┐
+     /─┤   ├─       /─┤   ├──□──┤   ├──□──┤   ├──□──┤   ├
+       └───┘          └───┘     └───┘     └───┘     └───┘
+
+    The number of CX gates generated with the decomposition without optimizations is,
+
+    .. math::
+
+        \frac{9}{16} 4^n - frac{3}{2} 2^n
+
+    If opt_a1=True, the CX count is further reduced by,
+
+    .. math::
+
+        \frac{1}{3} 4^{n - 2} - 1
+
+    If opt_a2=True, the CX count is further reduced by,
+
+    .. math::
+
+        4^{n - 1} - 1
+
+    Arguments:
+       mat (ndarray): unitary matrix to decompose
+       opt_a1 (bool): whether to try optimization A.1 from Shende. This should elliminate 1 cnot
+          per call. If True CZ gates are left in the output. If desired these can be further decomposed
+          to CX.
+       opt_a2 (bool): whether to try optimization A.2 from Shende. Not Implemented
+       decomposer_1q (None or Object): optional 1Q decomposer.
+       decomposer_2q (None or Object): optional 2Q decomposer.
+
+    Raises:
+       NotImplementedError: Occurs if opt_a2=True.
+
+    Return:
+       QuantumCircuit: Decomposed quantum circuit.
+    """
+    dim = mat.shape[0]
+    nqubits = int(np.log2(dim))
+    if opt_a2:
+        raise NotImplementedError("Optimization A.2 is not currently implemented.")
+    if np.allclose(np.identity(dim), mat):
+        return QuantumCircuit(nqubits)
+    if dim == 2:
+        if decomposer_1q is None:
+            decomposer_1q = one_qubit_decompose.OneQubitEulerDecomposer()
+        circ = decomposer_1q(mat)
+    elif dim == 4:
+        if decomposer_2q is None:
+            decomposer_2q = two_qubit_decompose.TwoQubitBasisDecomposer(CXGate())
+        circ = decomposer_2q(mat)
+    else:
```

Comment:
Actually this decomposition method will go down to single qubit unitaries but it's just that the known 3-cnot decomposition will yield lower cnot counts. If we let it go down to single qubits, this would correspond to the "QSD, l=1" column of the table above. I wouldn't raise, however; since you'd raise every time you tried to do this decomposition due to the recursion. I could see, however allowing it to use this the QSD all the way down to 1q but I wasn't sure how best to implement that or if anyone would actually prefer the higher CX cost.